patches/api/0440-Add-predicate-for-block-data-when-raytracing-blocks.patch

```diff
@@ -0,0 +1,90 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: TonytheMacaroni <tonythemacaroni123@gmail.com>
+Date: Wed, 6 Sep 2023 19:24:53 -0400
+Subject: [PATCH] Add predicate for block data when raytracing blocks
+
+
+diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
+index 6917931966377c51db88a3364997a110dd987970..8cd9b0c31bfd64c5e17ec6187c102c7c95d99599 100644
+--- a/src/main/java/org/bukkit/World.java
++++ b/src/main/java/org/bukkit/World.java
+@@ -1776,6 +1776,35 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
+     @Nullable
+     public RayTraceResult rayTraceBlocks(@NotNull Location start, @NotNull Vector direction, double maxDistance, @NotNull FluidCollisionMode fluidCollisionMode, boolean ignorePassableBlocks);
+ 
++    // Paper start
++    /**
++     * Performs a ray trace that checks for block collisions using the blocks'
++     * precise collision shapes.
++     * <p>
++     * If collisions with passable blocks are ignored, fluid collisions are
++     * ignored as well regardless of the fluid collision mode.
++     * <p>
++     * Portal blocks are only considered passable if the ray starts within
++     * them. Apart from that collisions with portal blocks will be considered
++     * even if collisions with passable blocks are otherwise ignored.
++     * <p>
++     * This may cause loading of chunks! Some implementations may impose
++     * artificial restrictions on the maximum distance.
++     *
++     * @param start the start location
++     * @param direction the ray direction
++     * @param maxDistance the maximum distance
++     * @param fluidCollisionMode the fluid collision mode
++     * @param ignorePassableBlocks whether to ignore passable but collidable
++     *     blocks (ex. tall grass, signs, fluids, ..)
++     * @param ignoreFilter predicate for block data to ignore when passing
++     *     through, or <code>null</code> to consider all blocks
++     * @return the ray trace hit result, or <code>null</code> if there is no hit
++     */
++    @Nullable
++    public RayTraceResult rayTraceBlocks(@NotNull Location start, @NotNull Vector direction, double maxDistance, @NotNull FluidCollisionMode fluidCollisionMode, boolean ignorePassableBlocks, @Nullable Predicate<BlockData> ignoreFilter);
```

Comment:
Both of these methods on World should use Position instead of Location as a world isn't required.

Reply:
Just check if the position is an instance of Location and then run those checks. A `checkFinite` method for Position might be good, but shouldn't be done in this PR.

Reply:
Should more overloads be added that take a `Position` instead of a `Location` for the pre-existing ray trace methods? Would require an additional one to call `World#rayTraceEntities` in `World#rayTrace` with a `Position`, and was wondering if it's necessary to add overloads for the shorter variants.

Reply:
At some point we will probably add some bytecode rewriting to just allow us to change those Location params to Position since Location implements Position. But that is for another time. For now, just use Position in new stuff where it is applicable.

Reply:
Alright, added an overload for `World#rayTraceEntities`, and changed the two added methods to take a `Position` .